src/main/java/org/cryptomator/linux/tray/AppindicatorTrayMenuController.java

```diff
@@ -0,0 +1,110 @@
+
+	private static final Logger LOG = LoggerFactory.getLogger(AppindicatorTrayMenuController.class);
+
+	private final SegmentScope scope = SegmentScope.auto();
+	private MemorySegment indicator;
+	private MemorySegment menu = gtk_menu_new();
+
+	@CheckAvailability
+	public static boolean isAvailable() {
+		return SystemUtils.IS_OS_LINUX && MemoryAllocator.isLoadedNativeLib();
+	}
+
+	@Override
+	public void showTrayIcon(URI uri, Runnable runnable, String s) throws TrayMenuException {
+		indicator = app_indicator_new(MemoryAllocator.ALLOCATE_FOR("org.cryptomator.Cryptomator"),
+				MemoryAllocator.ALLOCATE_FOR(getAbsolutePath(getPathString(uri))),
+				APP_INDICATOR_CATEGORY_APPLICATION_STATUS());
+		gtk_widget_show_all(menu);
+		app_indicator_set_menu(indicator, menu);
+		app_indicator_set_status(indicator, APP_INDICATOR_STATUS_ACTIVE());
+	}
+
+	@Override
+	public void updateTrayIcon(URI uri) {
+		app_indicator_set_icon(indicator, MemoryAllocator.ALLOCATE_FOR(getAbsolutePath(getPathString(uri))));
+	}
+
+	@Override
+	public void updateTrayMenu(List<TrayMenuItem> items) throws TrayMenuException {
+		menu = gtk_menu_new();
+		addChildren(menu, items);
+		gtk_widget_show_all(menu);
+		app_indicator_set_menu(indicator, menu);
+	}
+
+	@Override
+	public void onBeforeOpenMenu(Runnable runnable) {
+
+	}
+
+	private void addChildren(MemorySegment menu, List<TrayMenuItem> items) {
+		for (var item : items) {
+			// TODO: use Pattern Matching for switch, once available
+			if (item instanceof ActionItem a) {
+				var gtkMenuItem = gtk_menu_item_new();
+				gtk_menu_item_set_label(gtkMenuItem, MemoryAllocator.ALLOCATE_FOR(a.title()));
+				g_signal_connect_object(gtkMenuItem,
+						MemoryAllocator.ALLOCATE_FOR("activate"),
+						MemoryAllocator.ALLOCATE_CALLBACK_FOR(new ActionItemCallback(a), scope),
+						menu,
+						0);
+				gtk_menu_shell_append(menu, gtkMenuItem);
+			} else if (item instanceof SeparatorItem) {
+				var gtkSeparator = gtk_menu_item_new();
+				gtk_menu_shell_append(menu, gtkSeparator);
+			} else if (item instanceof SubMenuItem s) {
+				var gtkMenuItem = gtk_menu_item_new();
+				var gtkSubmenu = gtk_menu_new();
+				gtk_menu_item_set_label(gtkMenuItem, MemoryAllocator.ALLOCATE_FOR(s.title()));
+				addChildren(gtkSubmenu, s.items());
+				gtk_menu_item_set_submenu(gtkMenuItem, gtkSubmenu);
+				gtk_menu_shell_append(menu, gtkMenuItem);
+			}
+			gtk_widget_show_all(menu);
+		}
+	}
+	private String getAbsolutePath(String iconName) {
+		var res = getClass().getClassLoader().getResource(iconName);
+		if (null == res) {
+			throw new IllegalArgumentException("Icon '" + iconName + "' cannot be found in resource folder");
+		}
+		File file = null;
+		try {
+			file = Paths.get(res.toURI()).toFile();
+		} catch (URISyntaxException e) {
+			throw new IllegalArgumentException("Icon '" + iconName + "' cannot be converted to file", e);
+		}
+		return file.getAbsolutePath();
+	}
```

Comment:
> Does this really work? After all the icon is bundled within a jar.

It worked in the IDE. The final application is probably different than the compiled code in the IDE. Needs more testing.

> Is it really a "file" that can be read from within GTK?

No, the method takes the "icon_name" which is either the (short-) name of an icon installed on the system or its absolute path (what I am using here). The solution might be to use `app_indicator_set_icon_theme_path`. [libappindicator ](https://bazaar.launchpad.net/~indicator-applet-developers/libappindicator/trunk/view/head:/src/app-indicator.c)does not offer to much opportunities, though.

Reply:
> It worked in the IDE. The final application is probably different than the compiled code in the IDE. Needs more testing.

That's what I thought. In the IDE it is a "normal" file, it is not yet packed into a .jar archive. So either we need to unpack it from the jar to a temporary directory or bundle the icon with the installer in a way such that the icon will be installed as a regular file that we know the path of.

Reply:
Do you know how to resolve an icon "by name"? How do you register icon files? Will GTK use [this freedesktop specification](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html)? In that case we might just use the icon named `org.cryptomator.Cryptomator`.  The [desktop menu](https://github.com/cryptomator/cryptomator/blob/aa03bd119a3727cb850067d4034cfb4825d7eb88/dist/linux/common/org.cryptomator.Cryptomator.desktop#L5) makes use of this icon name, too.

Our [.deb file (ppa)](https://github.com/cryptomator/cryptomator/blob/aa03bd119a3727cb850067d4034cfb4825d7eb88/dist/linux/debian/cryptomator.install#L3-L5C79), the [Flatpak package](https://github.com/flathub/org.cryptomator.Cryptomator/blob/9d5f31a4e4f6cb3dfd147edd892e2f7c3c1e19b9/org.cryptomator.Cryptomator.yaml#L126-L128) as well as [our AppImage](https://github.com/cryptomator/cryptomator/blob/aa03bd119a3727cb850067d4034cfb4825d7eb88/dist/linux/appimage/build.sh#L68-L70) adhere to this standard.

@SailReal How about our AUR release?

Reply:
The `getAbsolutePath(...)` stuff was removed. As not needed any more.

We need to add a "org.cryptomator.Cryptomator-unlocked" SVG to our releases, so that it gets installed as an icon on Linux. The file is already in the repo as `src/main/resources/tray_icon_unlocked.svg` and can be removed afterwards, together with `tray_icon.svg` (which is already available as "org.cryptomator.Cryptomator").

Reply:
Done as well. For ppa, appimage and flatpak.

Reply:
Just for the record: Did you confirm, this works? Or did you find any source confirming that libappindicator and libayatana refer to this freedesktop spec with this "icon name"?

Reply:
In the first link [to the debian wiki](https://wiki.ubuntu.com/DesktopExperienceTeam/ApplicationIndicators) you provided the lib usees the freedesktop icon spec and libayatana is the same and that's what I did during testing all the time. So: yes.

Please note, I didn't build the ppa, nor the appimage or the flatpak app. The flatpak app requires the icon in the downloaded package that wasn't included at that time.